Approximate_solvers.ipynb

```diff
@@ -47,55 +92,74 @@
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
     "## Two-wave solvers \n",
-    "Since the Riemann solution impacts the overall numerical solution only based on how it modifies the two neighboring solution values, it seems reasonable to consider approximations in which only a single wave propagates in each direction.  The solution will have a single intermediate state $q_m$ such that ${\\cal W}_1 = q_m - q_l$ and ${\\cal W}_2 = q_r-q_m$.  There are apparently $m+2$ values to be determined: the middle state $q_m$ and the speeds $s_1, s_2$.  In order for the method to be conservative, it must satisfy (\\ref{adqdf}), and hence the $m$ conditions  \n",
+    "Since the Riemann solution impacts the overall numerical solution only based on how it modifies the two neighboring solution values, it seems reasonable to consider approximations in which only a single wave propagates in each direction.  The solution will have a single intermediate state $q_m$ such that ${\\cal W}_1 = q_m - q_\\ell$ and ${\\cal W}_2 = q_r-q_m$.  There are apparently $m+2$ values to be determined: the middle state $q_m \\in {\\mathbb R}^m$ and the speeds $s_1, s_2$.  In order for the approximate solver to be conservative, it must satisfy (\\ref{adqdf}), and hence the $m$ conditions  \n",
     "\\begin{align}\n",
-    "f(q_r) - f(q_l) = s_1 {\\cal W}_1 + s_2 {\\cal W}_2.\n",
+    "f(q_r) - f(q_\\ell) = s_1 {\\cal W}_1 + s_2 {\\cal W}_2.\n",
     "\\end{align}  \n",
     "This can be solved for the middle state to find  \n",
     "\\begin{align}  \\label{AS:middle_state}\n",
-    "q_m = \\frac{f(q_r) - f(q_l) - s_2 q_r + s_1 q_l}{s_1 - s_2}.\n",
+    "q_m = \\frac{f(q_r) - f(q_\\ell) - s_2 q_r + s_1 q_\\ell}{s_1 - s_2}.\n",
     "\\end{align}  \n",
     "It remains only to specify the wave speeds, and it is in this specification that the various two-wave solvers differ.  In the following sections we briefly discuss the choice of wave speed for a scalar problem; the choice for systems will be elaborated in subsequent chapters.\n",
     "\n",
-    "In addition to the names and references provided below, let us mention that this class of solvers is an ingredient in the so-called *central schemes*.  Due to the extreme simplicity of two-wave solvers, the resulting central schemes are often even referred to as being \"Riemann-solver-free\"."
+    "Typically $s_1 < 0 < s_2$ and so the intermediate state we need is $Q^* = q_m$ and $F^* = f(Q^*)$.  However, in some cases both $s_1$ and $s_2$ could have the same sign, in which case $F^*$ is either $f(q_\\ell)$ or $f(q_r)$.\n",
```

Comment:
Are there approximate two-wave solvers where this can actually occur?  The ones I can think of assume the waves travel in opposite directions.

Reply:
If you apply HLL to a supersonic Riemann problem for Euler (or supercritical for SWE) then the eigenvalues could all have the same sign in both states so the formula at the bottom of this notebook would give the same sign for both speeds.

Reply:
I agree that the wave speeds in the true Riemann solution can all have the same sign. But I don't know of an approximate two-wave solver in the literature that allows two waves with the same sign.  The formula at the bottom of this notebook is for a scalar problem, not for systems.  From my understanding of the HLL solver, in the supersonic/supercritical case, the other wave would be given speed zero.

Reply:
In my mind, the idea of a two wave solver is to capture all the effects of waves going in a given direction through a single consolidated wave (basically a fluctuation).  Of course, one could design a different kind of two-wave solver; I just haven't seen it.

Reply:
I pointed to the wrong scalar equations, but see the expressions for $s_1$ and $s_2$ in `Shallow_water_approximate` or `Euler_approximate`.  Won't these both be the same sign in the supercritical cases?

Reply:
Hmmmm, you're right. I guess this is a misconception on my part, although I'm certain I have seen some papers where they take the min/max with zero for those speeds.  Anyway, I withdraw my objection to this.